src/main/scala/org/apache/spark/rdd/OrderedRDD.scala

```diff
@@ -8,105 +8,112 @@ import scala.collection.mutable.ArrayBuffer
-
-      val minMax = arr.map(x => (x._3, x._4))
-      val sortedBetweenPartitions = minMax.tail.map(_._1).zip(minMax.map(_._2))
-        .forall { case (nextMin, lastMax) => tOrd.gt(nextMin, lastMax) }
-
-      if (sortedBetweenPartitions) {
-        val ab = mutable.ArrayBuilder.make[T]()
-        minMax.take(minMax.length - 1).foreach { case (_, max) => ab += max }
-
-        val partitioner = OrderedPartitioner[T, K](ab.result(), projectKey)
-        assert(partitioner.numPartitions == rdd.partitions.length)
-
-        if (allKSorted) {
+    if (rdd.partitions.isEmpty)
+      return empty(rdd.sparkContext, projectKey)
+
+    val keys = fastKeys.getOrElse(rdd.map(_._1))
+
+    val keyInfoOption = anyFailAllFail[Array, PartitionKeyInfo[T]](
+      keys.mapPartitionsWithIndex { case (i, it) =>
+        Iterator(if (it.hasNext)
+          Some(PartitionKeyInfo.apply(i, projectKey, it))
+        else
+          None)
+      }.collect())
+
+    val partitionsSorted =
+      keyInfoOption.exists(keyInfo =>
+        keyInfo.tail.zip(keyInfo).forall { case (pi1, pi2) =>
+          tOrd.lt(pi1.max, pi2.min)
+        })
+
+    if (partitionsSorted) {
+      val keyInfo = keyInfoOption.get
+      val partitioner = OrderedPartitioner[T, K](keyInfo.init.map(pi => pi.max), projectKey)
+      val sortedness = keyInfo.map(_.sortedness).min
+      (sortedness: @unchecked) match {
+        case PartitionKeyInfo.KSORTED =>
+          assert(sortedness == PartitionKeyInfo.KSORTED)
           info("Coerced sorted dataset")
-          Some(new OrderedRDD[T, K, V](rdd, partitioner))
-        } else {
-          val sortedRDD = rdd.mapPartitionsWithIndex { case (i, it) =>
-            if (arr(i)._2)
-              it
-            else it.localKeySort[T](projectKey)
-          }
-          info("Coerced almost-sorted dataset")
-          Some(new OrderedRDD(sortedRDD, partitioner))
-        }
-      }
-      else None
-    }
+          new OrderedRDD(rdd, partitioner)
 
-    def verifySortedness(i: Int, it: Iterator[K]): Iterator[Option[(Int, Boolean, T, T)]] = {
-      if (it.isEmpty)
-        Iterator(None)
-      else {
-        var sortedT = true
-        var sortedK = true
-        var continue = true
-        val firstK = it.next()
-        var lastK = firstK
-        var lastT = projectKey(firstK)
-
-        while (it.hasNext && continue) {
-          val k = it.next()
-          val t = projectKey(k)
-          if (tOrd.lt(t, lastT)) {
-            sortedT = false
-            continue = false
-          } else if (kOrd.lt(k, lastK))
-            sortedK = false
-
-          lastK = k
-          lastT = t
-        }
-        if (!sortedT)
-          Iterator(None)
-        else Iterator(Some((i, sortedK, projectKey(firstK), projectKey(lastK))))
+        case PartitionKeyInfo.TSORTED =>
+          info("Coerced almost-sorted dataset")
+          new OrderedRDD(rdd.mapPartitions { it =>
+            it.localKeySort(projectKey)
```

Comment:
This may not matter, but previously we were using the partition summaries to ask whether a partition needs to be sorted (in a T-sorted K-unsorted dataset, most partitions are probably K-sorted).  If the localKeySort is basically free, this is fine.